### PR TITLE
fix(components): Update types for GCS components so it's compatible with v2.

### PR DESCRIPTION
**Pull request by neuromage**
Explicitly annotate URI types as String so they get treated
appropriately in v2.

**Checklist:**
- [X] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

**Comment by Ark-kun** [comment]
It looks like this change might break existing user and customer pipelines.
We are aware that there are customers that use these components. Changing input and output types will cause the pipelines to start failing dues to incompatible types.

Also, changing a handful of components in this repo won't change the components in the user repositories.

Perhaps this issue can be fixed on the compiler side.
WDYT?

File: components/google-cloud/storage/download/component.yaml [hunk 1]
```diff
@@ -1,8 +1,8 @@
 name: Download from GCS
 inputs:
-- {name: GCS path, type: URI}
+- {name: GCS path, type: String}
```
Comment (Ark-kun) [inline components/google-cloud/storage/download/component.yaml:3]: This change makes the component incompatible with typed component outputs or pipeline parameters.

File: components/google-cloud/storage/download/component.yaml [hunk 2]
```diff
@@ -1,8 +1,8 @@
...
 outputs:
-- {name: Data}
+- {name: Data, type: Artifact}
```
Comment (Ark-kun) [inline components/google-cloud/storage/download/component.yaml:5]: This change makes the component incompatible with most downstream components. 
Type `Artifact` is not compatible with the type `XGBoostModel`, `PyTorchModel` etc.

Comment (neuromage) [reply components/google-cloud/storage/download/component.yaml]: Ok, I reverted this change.

**Comment by Bobgy** [comment]
Shall we enable this example in the same PR to verify?
https://github.com/kubeflow/pipelines/blob/f757a8842fd6d1ed3d37ab38c0707d127ea2048c/samples/core/exit_handler/exit_handler_test.py#L24-L34
https://github.com/kubeflow/pipelines/blob/f757a8842fd6d1ed3d37ab38c0707d127ea2048c/samples/test/legacy_exit_handler_test.py#L32-L42

**Comment by neuromage** [comment]
@Bobgy if the samples are meant to load the components from the URI, then it's better to submit this change before enabling them.

**Comment by Bobgy** [comment]
Good point, I missed that.
I'd say they are not designed for loading by URL, we can also load from relative path like https://github.com/kubeflow/pipelines/blob/d362cab2046b92efa5f015c6a980bc9371b4b63b/samples/core/visualization/tensorboard_minio.py#L20-L29

**Comment by Bobgy** [comment]
Another risk is that, sample pipelines get their component version bumped automatically during release, we can only know how many sample pipelines break after the change after next RC release.

**Comment by Bobgy** [comment]
Another idea, what do you think about copying components to v2 folder? because this is basically a breaking change

**Comment by Ark-kun** [comment]
> Another idea, what do you think about copying components to v2 folder? because this is basically a breaking change

I'm not sure about this. The components in this PR are still v1-compatible.

I'm OK with changing the component input types when it is warranted (for example changing `Text` to `JsonArray` for better SDK support).
However in this case I wonder whether we should fix our compiler instead of asking all users to change their components: Here is a PR that fixes the issue on the compiler side without a need to change user components: https://github.com/kubeflow/pipelines/pull/5478

**Comment by Bobgy** [comment]
For similar reasons as Chen, I don't like some hacky natures of #5478 either.
I think we should now think about documenting the migration path for v2 compatible mode, because it clearly isn't compatible with all v1 pipelines either unlike initially planned.

/lgtm

**Comment by neuromage** [comment]
> Another idea, what do you think about copying components to v2 folder? because this is basically a breaking change
> 
> EDIT: after rethinking over this, copying probably don't make sense, because we want people to migrate to the new semantics, also these components still make sense for KFP v1. The only thing it breaks is the type ontology KFP v1 created via these components and samples, now people need to relearn that parameters must be typed String.

+1. I think there are two steps we should take:
- send a big clean up PR like this one for all our components and samples to streamline the types everywhere. This will encourage users to start writing component definitions in this way.
- update the type-checking code in KFP SDK, so `Artifact` is like an `Any`-typed artifact. That's useful for generic components such as the one here. Right now, the empty type is treated as `Any` by compiler, but this is not a documented feature.

**Pull request merged**